rockcraft/oci.py

```diff
@@ -189,6 +200,83 @@ def add_layer(
         name = self.image_name.split(":", 1)[0]
         return self.__class__(image_name=f"{name}:{tag}", path=self.path)
 
+    def add_user(
+        self,
+        prime_dir: Path,
+        base_layer_dir: Path,
+        tag: str,
+        username: str,
+        uid: int,
+    ) -> None:
+        """Create a new ROCK user.
+
+        :param prime_dir: Path to the user-defined parts' primed content.
+        :param base_layer_dir: Path to the base layer's root filesystem.
+        :param tag: The ROCK's image tag.
+        :param username: Username to be created. Same as group name.
+        :param uid: UID of the username to be created. Same as GID.
+        """
+        # pylint: disable=too-many-arguments
+        user_files = {"passwd": "", "group": "", "shadow": ""}
+
+        prime_dir_etc = prime_dir / "etc"
+        base_layer_dir_etc = base_layer_dir / "etc"
+        # Being cautious about possible changes (edits or removals) in
+        # /etc/{passwd,group,shadow} done by the user through overlay scripts.
+        # Basically:
+        #  - if it exists in prime, use it,
+        #  - if it doesn't exist in prime AND isn't "whiteout", use the base,
+        #  - if it is "whiteout" or doesn't exist anywhere, use an empty file.
+        # NOTE: "shadow" is only modified it it already exists.
+        for u_file in user_files:
+            if (prime_dir_etc / u_file).exists():
+                user_files[u_file] = (prime_dir_etc / u_file).read_text()
+            elif (base_layer_dir_etc / u_file).exists() and not (
+                prime_dir_etc / f".wh.{u_file}"
+            ).exists():
+                user_files[u_file] = (base_layer_dir_etc / u_file).read_text()
+
+        if (
+            f"{username}:" in user_files["passwd"]
```

Comment:
Do we care about the edge case of an existing username that ends with the added username?

e.g. adding the user `daemon` but the passwd file already contains a `somedaemon` user?

(Genuine question and I presume not, but if we do we might want to ensure that the username immediately follows a newline.)

Reply:
nice catch. that's a good corner case to cover indeed. I've improved the conflict checking in https://github.com/canonical/rockcraft/pull/233/commits/e9db7fbfc6c6615e2961636f4c47bfe455b204ab